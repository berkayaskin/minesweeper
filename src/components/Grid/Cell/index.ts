export { Cell, ClosedFrame, isActiveCell } from './Cell'
export type { CellProps } from './Cell'
