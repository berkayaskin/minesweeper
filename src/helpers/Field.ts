export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
export type Field = Cell[][]
export type Coords = [number, number]

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  mark: 11,
  weakMark: 12,
}
