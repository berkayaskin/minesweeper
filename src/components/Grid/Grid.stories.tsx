import { Meta, Story } from '@storybook/react'

import { Grid, GridProps } from './Grid'

export default {
  title: 'Grid/Grid',
  component: Grid,
} as Meta

const Template: Story<GridProps> = (args) => <Grid {...args} />

export const GridExample = Template.bind({})
GridExample.args = {
  children: [
    [9, 2, 9, 1, 0, 0, 1, 1, 1, 1, 1, 1],
    [1, 2, 2, 2, 1, 0, 1, 9, 1, 1, 9, 1],
    [0, 0, 1, 9, 10, 0, 2, 2, 2, 1, 1, 1],
    [0, 0, 10, 10, 1, 0, 1, 9, 1, 2, 2, 2],
    [0, 1, 1, 2, 2, 9, 1, 1, 1, 0, 0, 0],
    [0, 1, 9, 3, 9, 2, 10, 0, 0, 2, 1, 1],
    [0, 2, 2, 4, 9, 2, 10, 1, 1, 1, 9, 1],
    [0, 1, 9, 2, 1, 1, 1, 9, 1, 2, 2, 2],
    [0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
    [0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
    [0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
    [0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
  ],
}
