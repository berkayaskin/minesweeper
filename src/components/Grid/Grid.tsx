import { Coordinates, Field } from '@/helpers/Field'
import styled from '@emotion/styled'
import { Cell } from './Cell'

export interface GridProps {
  /**
   * Field data
   */
  children: Field
  /**
   * onClick handler
   * @param coordinates
   */
  onClick: (coordinates: Coordinates) => void
  /**
   * onContextMenu handler
   * @param coordinates
   */
  onContextMenu: (coordinates: Coordinates) => void
}

export const Grid = ({ children, ...rest }: GridProps) => (
  <Wrapper size={children.length}>
    {children.map((row, y) =>
      row.map((cell, x) => (
        <Cell key={`${cell}_${x}_${y}`} coordinates={[x, y]} {...rest}>
          {cell}
        </Cell>
      )),
    )}
  </Wrapper>
)

interface WrapperProps {
  size: number
}

const Wrapper = styled.div<WrapperProps>`
  display: grid;
  grid-template-columns: repeat(${({ size }) => size}, auto);
  width: max-content;
  padding: 1vw;
`
