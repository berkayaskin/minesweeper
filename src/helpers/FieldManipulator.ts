import { CellState, Coordinates, Field } from './Field'

export const getNeighbours = ([x, y]: Coordinates): Record<string, [number, number]> => {
  return {
    topLeft: [x - 1, y - 1],
    top: [x - 1, y],
    topRight: [x - 1, y + 1],
    left: [x, y - 1],
    right: [x, y + 1],
    bottomLeft: [x + 1, y - 1],
    bottom: [x + 1, y],
    bottomRight: [x + 1, y + 1],
  }
}

export const isItemExistInField = ({ length }: Field, [x, y]: Coordinates): boolean => {
  return x >= 0 && x < length && y >= 0 && y < length
}

export const incrementNeighbours = (field: Field, coordinates: Coordinates): Field => {
  const neighbours = getNeighbours(coordinates)
  const newField = [...field]

  Object.values(neighbours).forEach(([x, y]) => {
    if (isItemExistInField(field, [x, y])) {
      if (newField[x][y] !== CellState.bomb) {
        newField[x][y]++
      }
    }
  })

  return newField
}
