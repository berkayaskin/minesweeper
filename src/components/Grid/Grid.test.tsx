import { Field } from '@/helpers/Field'
import { render } from '@testing-library/react'
import { Grid } from './Grid'

export const MockFieldData: Field = [
  [9, 2, 9, 1, 0, 0, 1, 1, 1, 1, 1, 1],
  [1, 2, 2, 2, 1, 0, 1, 9, 1, 1, 9, 1],
  [0, 0, 1, 9, 10, 0, 2, 2, 2, 1, 1, 1],
  [0, 0, 10, 10, 1, 0, 1, 9, 1, 2, 2, 2],
  [0, 1, 1, 2, 2, 9, 1, 1, 1, 0, 0, 0],
  [0, 1, 9, 3, 9, 2, 10, 0, 0, 2, 1, 1],
  [0, 2, 2, 4, 9, 2, 10, 1, 1, 1, 9, 1],
  [0, 1, 9, 2, 1, 1, 1, 9, 1, 2, 2, 2],
  [0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
  [0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
  [0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
  [0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
]

it('should render grid correctly', () => {
  const props = {
    onClick: jest.fn(),
    onContextMenu: jest.fn(),
  }

  const { asFragment } = render(<Grid {...props}>{MockFieldData}</Grid>)
  expect(asFragment()).toMatchSnapshot()
})
