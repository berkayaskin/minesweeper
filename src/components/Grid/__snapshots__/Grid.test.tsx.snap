// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render grid correctly 1`] = `
<DocumentFragment>
  <div
    class="css-2gtdmb"
    size="12"
  >
    <div
      class="css-1na1t6s"
      data-testid="9_0,0"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_1,0"
    >
      2
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_2,0"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_3,0"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_4,0"
    />
    <div
      class="css-1bascby"
      data-testid="0_5,0"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,0"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_7,0"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,0"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,0"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_10,0"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,0"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_0,1"
    >
      1
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_1,1"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_2,1"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_3,1"
    >
      2
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_4,1"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_5,1"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,1"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_7,1"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,1"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,1"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_10,1"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,1"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,2"
    />
    <div
      class="css-1bascby"
      data-testid="0_1,2"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_2,2"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_3,2"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-rqpdce"
      data-testid="10_4,2"
    />
    <div
      class="css-1bascby"
      data-testid="0_5,2"
    />
    <div
      class="css-1w4pte2"
      data-testid="2_6,2"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_7,2"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_8,2"
    >
      2
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,2"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_10,2"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,2"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,3"
    />
    <div
      class="css-1bascby"
      data-testid="0_1,3"
    />
    <div
      class="css-rqpdce"
      data-testid="10_2,3"
    />
    <div
      class="css-rqpdce"
      data-testid="10_3,3"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_4,3"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_5,3"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,3"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_7,3"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,3"
    >
      1
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_9,3"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_10,3"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_11,3"
    >
      2
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,4"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,4"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_2,4"
    >
      1
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_3,4"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_4,4"
    >
      2
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_5,4"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_6,4"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_7,4"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,4"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_9,4"
    />
    <div
      class="css-1bascby"
      data-testid="0_10,4"
    />
    <div
      class="css-1bascby"
      data-testid="0_11,4"
    />
    <div
      class="css-1bascby"
      data-testid="0_0,5"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,5"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_2,5"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1e2a6pr"
      data-testid="3_3,5"
    >
      3
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_4,5"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_5,5"
    >
      2
    </div>
    <div
      class="css-rqpdce"
      data-testid="10_6,5"
    />
    <div
      class="css-1bascby"
      data-testid="0_7,5"
    />
    <div
      class="css-1bascby"
      data-testid="0_8,5"
    />
    <div
      class="css-1w4pte2"
      data-testid="2_9,5"
    >
      2
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_10,5"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,5"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,6"
    />
    <div
      class="css-1w4pte2"
      data-testid="2_1,6"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_2,6"
    >
      2
    </div>
    <div
      class="css-1qy3cpa"
      data-testid="4_3,6"
    >
      4
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_4,6"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_5,6"
    >
      2
    </div>
    <div
      class="css-rqpdce"
      data-testid="10_6,6"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_7,6"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,6"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,6"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_10,6"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,6"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,7"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,7"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_2,7"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_3,7"
    >
      2
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_4,7"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_5,7"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_6,7"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_7,7"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,7"
    >
      1
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_9,7"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_10,7"
    >
      2
    </div>
    <div
      class="css-1w4pte2"
      data-testid="2_11,7"
    >
      2
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,8"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,8"
    >
      1
    </div>
    <div
      class="css-rqpdce"
      data-testid="10_2,8"
    />
    <div
      class="css-rqpdce"
      data-testid="10_3,8"
    />
    <div
      class="css-1bascby"
      data-testid="0_4,8"
    />
    <div
      class="css-1bascby"
      data-testid="0_5,8"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,8"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_7,8"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,8"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,8"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_10,8"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,8"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,9"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,9"
    >
      1
    </div>
    <div
      class="css-rqpdce"
      data-testid="10_2,9"
    />
    <div
      class="css-rqpdce"
      data-testid="10_3,9"
    />
    <div
      class="css-1bascby"
      data-testid="0_4,9"
    />
    <div
      class="css-1bascby"
      data-testid="0_5,9"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,9"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_7,9"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,9"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,9"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_10,9"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,9"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,10"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,10"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_2,10"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_3,10"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_4,10"
    />
    <div
      class="css-1bascby"
      data-testid="0_5,10"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,10"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_7,10"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,10"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,10"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_10,10"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,10"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_0,11"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_1,11"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_2,11"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_3,11"
    >
      1
    </div>
    <div
      class="css-1bascby"
      data-testid="0_4,11"
    />
    <div
      class="css-1bascby"
      data-testid="0_5,11"
    />
    <div
      class="css-1yu2ia5"
      data-testid="1_6,11"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_7,11"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_8,11"
    >
      1
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_9,11"
    >
      1
    </div>
    <div
      class="css-1na1t6s"
      data-testid="9_10,11"
    >
      <div
        class="css-b14p9c"
      />
    </div>
    <div
      class="css-1yu2ia5"
      data-testid="1_11,11"
    >
      1
    </div>
  </div>
</DocumentFragment>
`;
